go-controller/pkg/node/udn_isolation.go

```diff
@@ -293,52 +295,77 @@ func (m *UDNHostIsolationManager) updateKubeletCgroup() error {
 // If a new cgroup is created, you load the filtering policy for the new cgroup and then add
 // processes to that cgroup. You only have to follow the right sequence to avoid problems.
 func (m *UDNHostIsolationManager) runKubeletRestartTracker(ctx context.Context) (err error) {
-	conn, err := dbus.NewSystemdConnectionContext(ctx)
+
+	conn, err := dbus.Dial("unix:path=/run/systemd/private", dbus.WithContext(ctx))
 	if err != nil {
-		return fmt.Errorf("failed to connect to systemd: %w", err)
+		return err
 	}
+
 	defer func() {
 		if err != nil {
-			conn.Close()
+			if err := conn.Close(); err != nil {
+				klog.Errorf("Error closing dbus connection for UDN isolation: %v", err)
+			}
 		}
 	}()
 
-	err = conn.Subscribe()
+	// Only use EXTERNAL method, and hardcode the uid (not username)
+	// to avoid a username lookup (which requires a dynamically linked
+	// libc)
+	methods := []dbus.Auth{dbus.AuthExternal(strconv.Itoa(os.Getuid()))}
+
+	err = conn.Auth(methods)
 	if err != nil {
-		return fmt.Errorf("failed to subscribe to systemd events: %w", err)
+		return err
 	}
-	// interval is important here as we need to catch the restart state, before it is running again
-	events, errChan := conn.SubscribeUnitsCustom(50*time.Millisecond, 0, func(u1, u2 *dbus.UnitStatus) bool { return *u1 != *u2 },
-		func(s string) bool {
-			return s != "kubelet.service"
-		})
+
+	signalChan := make(chan *dbus.Signal, 10)
+	conn.Signal(signalChan)
+
 	// run until context is cancelled
 	go func() {
-		waitingForActive := false
 		for {
 			select {
 			case <-ctx.Done():
-				conn.Close()
+				if err := conn.Close(); err != nil {
+					klog.Errorf("Error closing dbus connection for UDN isolation: %v", err)
+				}
 				return
-			case event := <-events:
-				for _, status := range event {
-					if status.ActiveState != "active" {
-						waitingForActive = true
-					} else if waitingForActive {
-						klog.Infof("Kubelet was restarted, re-applying UDN host isolation")
-						err = m.updateKubeletCgroup()
-						if err != nil {
-							klog.Errorf("Failed to re-apply UDN host isolation: %v", err)
-						} else {
-							waitingForActive = false
+			case signal := <-signalChan:
+				klog.V(5).Infof("D-Bus event received: %#v", signal)
+				// Extract unit name from path
+				unitPath := signal.Path
+				parts := strings.Split(string(unitPath), "/")
+				if len(parts) < 6 || parts[4] != "unit" {
+					continue
+				}
+				escapedUnit := parts[5]
+				unitName := strings.ReplaceAll(escapedUnit, "_2e", ".")
+
+				if unitName == "kubelet.service" {
+					changes := signal.Body[1].(map[string]dbus.Variant)
+					if state, exists := changes["ActiveState"]; exists {
+						newState := state.Value().(string)
+						if newState == "active" {
```

Comment:
on my kind cluster, this causes 4 cgroup updates for every kubelet restart. That is why I had this `waitingForActive` logic.

Reply:
I see the same thing in my setup, but the cost of executing 4 times on a kubelet restart seems minimal when i pprof'ed.

Reply:
i mean why make 4 system calls when you can do 1? in case you miss deactivating events? is it likely to happen? or some other reason?

Reply:
exactly...i'm worried about missing events. I dont want to internally track state in ovnk. I think it could be error prone. Kubelet doesn't restart often so when it happens its not a big hit to just reprogram the nftables rules 4x. We can explore maybe examining the properties event in more detail to see if there is something else we could filter further on. I'll try to take a look at that, but I dont think this is a huge issue.